internal/expr/querybuilder.go

```diff
@@ -122,16 +114,28 @@ func (qb *queryBuilder) checkAllArgsUsed(typeToValue typeinfo.TypeToValue) error
 	return nil
 }
 
+// inputAssigner assigns input numbers to input expressions.
+type inputAssigner struct {
+	inputCount int
+}
```

Comment:
To me this looks like a code smell, that probably there is too much abstraction for the given complexity. In my opinion, what you had at https://github.com/canonical/sqlair/pull/145/commits/afade243d0d0a68049dccfe4167e009272f92f9a was great, it moved the logic to encapsulate it and made the code clearer and more self-contained.

We now seem to be marching towards more encapsulation and decoupling, which is a worthwhile goal don't get me wrong. However, to me the results look a little bit too complex. Maybe we could wait until we need to add something else to the code and then do the refactor once we have new concrete use-cases.

Reply:
As discussed in person, we'll agree to disagree. In my opinion this is not adding complexity, it is making explicit complexity that is already there. The `inputAssigner` allows the `bindInputs` method on the `insertColumn` to take care of assigning the input numbers for the `boundInsertColumn` to use. Before it was being done by the caller based on the return values which (in a very minor way) broke the encapsulation of the the `boundColumn`.

This also makes changes to the way inputs are assigned a lot easier. We just change the `inputAssigner` struct and then follow through the compiler errors where `assignInputs` is called.